qualtran/bloqs/rotations/rz_via_phase_gradient.py

```diff
@@ -0,0 +1,110 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    QBit,
+    QDType,
+    QUInt,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic.controlled_add_or_subtract import ControlledAddOrSubtract
+from qualtran.bloqs.bookkeeping import Cast
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@frozen
+class RzViaPhaseGradient(Bloq):
+    r"""Apply a controlled-Rz using a phase gradient state.
+
+    Implements the following unitary action:
+
+    $$
+        |\psi\rangle \otimes |x\rangle \mapsto \text{Rz}(x) |\psi\rangle \otimes |x\rangle
+    $$
+
+    for every state $|\psi\rangle$ and every $x$, or equivalently
+
+    $$
+        |b\rangle|x\rangle \mapsto |b\rangle e^{- (-1)^b i x/2} |x\rangle
+    $$
+
+    for every $b \in \{0, 1\}$ and every $x$.
+
+    Args:
+        angle_dtype: Data type for the `angle_data` register.
+        phasegrad_dtype: Data type for the phase gradient register.
+
+    Registers:
+        q: The qubit to apply Rz on.
+        angle: The rotation angle in radians.
+        phase_grad: The phase gradient register of sufficient width.
+
+    References:
+         [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](
```

Comment:
There should be no line break in the reference link I believe.

Reply:
Fixed